feat: custom DQL queries within catalog-info.yaml

blasget opened the pull request:
# Introduction

resolves #46 

## Technical Solution Before This PR

Previously, custom DQL queries had to be defined within the `app-config.yaml` file of the Backstage instance. This approach did not allow developers to define their own custom actions for their components. Moreover, for each custom query, an `EntityDqlQueryCard` with the query ID had to be added.

## Technical Solution After This PR

The `EntityCatalogInfoQueryCard` is used to display a result table for each query defined in the `spec.queries` property of an entity in the `catalog-info.yaml` file. This eliminates the need to provide query IDs.

To showcase custom queries, an additional route was added to the `EntityPage.tsx` file in the demo-backstage component.

Additionally, three sample queries were added to the demo-backstage component:
- Error Logs
- Problem Events
- Security Vulnerabilities

![image](https://github.com/Dynatrace/backstage-plugin/assets/119623079/bf06ba0b-1488-46a6-aa53-edc4d7d4987e)

#### :heavy_check_mark: Common checklist

- [x] Added tests for new functionality and regression tests for bug fixes
- [x] Screenshots of before and after attached (for UI changes)
- [x] Added or updated documentation (if applicable)
- [x] Helpful resources linked (if applicable)

Kirdock commented on plugins/dql-backend/src/service/router.ts:
```diff
@@ -53,6 +54,20 @@ export const createRouter = async (
   const router = Router();
   router.use(express.json());
 
+  router.get('/catalog/:queryId', async (req, res) => {
```
So at the moment there is a call for every query inside the component in `catalog-info.yaml`.
I'm not sure if this will lead to any side effects, except when the user has some rate-limits in place for their backstage instance.
@MrManny @TorbenGallobDynatrace your opinion on that?

TorbenGallobDynatrace replied:
Definitely the better solution.

The backend should return the results in a nested array, as the backend already knows which queries the component has.

johannes-b commented on README.md:
```diff
@@ -298,6 +298,47 @@ by its ID with the `custom.` prefix:
+  lifecycle: experimental
+  system: integrations
+  queries:
+    - name: Fetch all Davis events 1
```
Can we exchange this example with another one, for instance, the PROBLEM_EVENTS from below?

The reason is that I want to avoid confusion on what DAVIS Events are.

johannes-b commented on README.md:
````diff
@@ -298,6 +298,47 @@ by its ID with the `custom.` prefix:
+```
+
+As mentioned before, queries can contain placeholders. In the catalog-info.yaml
+file, the placeholders are prefixed with a single `$`. For example:
````
Instead of providing examples, please link to: https://github.com/Dynatrace/backstage-plugin/blob/eac6adfe0c25fc7a4e5b0b7d05d5dc83464f3652/README.md#custom-queries

The text should be: `Please find the supported placeholders listed [here](https://github.com/Dynatrace/backstage-plugin/blob/eac6adfe0c25fc7a4e5b0b7d05d5dc83464f3652/README.md#custom-queries).`

johannes-b commented on README.md:
```diff
@@ -298,6 +298,47 @@ by its ID with the `custom.` prefix:
+    - name: Fetch all Davis events 1
+      query: >
+        fetch events | filter event.kind == "DAVIS_EVENT" | fields event.kind,
+        timestamp
```
It would be great if the query example contains a placeholder.

johannes-b commented on catalog-info.yaml:
```diff
@@ -10,6 +10,33 @@ spec:
   owner: user:default/mjakl
   lifecycle: experimental
   system: integrations
+  queries:
```
Does it make sense to extend these three examples with a `filter` that contains a placeholder (e.g., componentName). This would avoid listing all error logs, problem events, and security vulnerabilities.

Kirdock commented on README.md:
```diff
@@ -298,6 +298,47 @@ by its ID with the `custom.` prefix:
+  owner: user:default/mjakl
+  lifecycle: experimental
+  system: integrations
+  queries:
```
Annotations only support key-value pairs as far as I know. So an array or an object is not possible

johannes-b replied:
Correct. Just looked it up here: https://backstage.io/docs/features/software-catalog/descriptor-format/#annotations-optional - Hence, this is no option. 

@blasget could you please do a quick research on how other plugins solve this?

blasget replied:
The interface does not allow to add a property eg. dynatrace.queries. 
![image](https://github.com/user-attachments/assets/b9d43878-e5f6-4727-9ab3-b9098a9e1970)

blasget commented:
# Updated PR

## Introduction

resolves #46 and
resolves #107 

## Technical Solution Before This PR

Previously, custom DQL queries had to be defined within the `app-config.yaml` file of the Backstage instance. This approach did not allow developers to define their own custom actions for their components. Moreover, for each custom query, an `EntityDqlQueryCard` with the query ID had to be added.

There was no default query to get site reliability guardian validations and display them in Backstage.

## Technical Solution After This PR

The `Entity` interface doesn’t support adding custom properties such as `dynatrace.queries`. As a result, the decision was made to use annotations and Dynatrace Notebooks instead. This method offers a significant benefit: it enables users to create and test their custom queries directly within their Dynatrace environment. These queries can be included in a backstage entity via annotations defined in the `catalog-info.yaml` file. There are two types of notebook annotations supported:

1. dynatrace.com/notebook-url:  https://my-environment.dynatrace.com/ui/apps/dynatrace.notebooks/notebook/my-notebook-id
2. dynatrace.com/notebook-id: my-notebook-id

![image](https://github.com/user-attachments/assets/904e0bab-a223-491c-8e2d-98e79d7c6b2b)

The `EntityNotebookQueriesCard` is used to display a result table for each query defined in a [Dynatrace Notebook](https://docs.dynatrace.com/docs/observe-and-explore/dashboards-and-notebooks/notebooks). Placeholders are not supported.

To showcase custom queries, an additional route was added to the `EntityPage.tsx` file in the demo-backstage component.

Additionally, three sample queries were added to the README.md:
- Error Logs
- Problem Events
- Security Vulnerabilities

A default query has been integrated to automatically fetch SRG validation results from Dynatrace.
Each entry in the validation results table now includes a direct link that opens the Site Reliability Guardian (SRG) view.

![image](https://github.com/user-attachments/assets/b5ab4943-05e0-40fe-bf79-41624b8b85a5)

#### :heavy_check_mark: Common checklist

- [x] Added tests for new functionality and regression tests for bug fixes
- [x] Screenshots of before and after are attached (for UI changes)
- [x] Added or updated documentation (if applicable)
- [x] Helpful resources linked (if applicable)

Kirdock commented on plugins/dql-backend/src/service/queries.ts:
```diff
@@ -87,4 +88,54 @@ export const dynatraceQueries: Record<
+    let filterString = '';
+
+    if (catalogTags) {
+      const tags = catalogTags.split(',').map(tag => {
+        const [key, value] = tag.split('=');
+        if (value === undefined) {
+          return { [key]: undefined };
+        }
+        return { [key]: value || '' };
+      });
+
+      if (tags.length > 0) {
+        filterString += '| filter ';
+        filterString += tags
+          .map(tag => {
+            const key = Object.keys(tag)[0];
+            const value = tag[key];
+            return value !== undefined
+              ? `in (tags[${key}], "${value}")`
+              : `isNotNull(tags[${key}])`;
+          })
```
Minor (no change needed): Could be improved a bit. Instead of `{[key]: value}` we could also write `{key, value}`, because we don't add more to the object and don't really need this object structure.
For the mapping function we could then just use `tag.key` and `tag.value`

Kirdock commented on plugins/dql-backend/src/service/queries.ts:
```diff
@@ -87,4 +88,54 @@ export const dynatraceQueries: Record<
+    }
+
+    return `
+    fetch bizevents, from: -2d
```
We should also discuss how we want to proceed with the timeframe. @johannes-b any opinions on a default timeframe for SRG validations?

Kirdock commented on plugins/dql-backend/src/service/queries.ts:
```diff
@@ -87,4 +88,54 @@ export const dynatraceQueries: Record<
+            const key = Object.keys(tag)[0];
+            const value = tag[key];
+            return value !== undefined
+              ? `in (tags[${key}], "${value}")`
```
To be on the safe side, we can escape the key, so that spaces in keys are allowed
```suggestion
              ? `in (tags[\`${key}\`], "${value}")`
```

Kirdock reviewed (commented):
First of all great work!

I would also have some questions :) 
- Which timeframe is taken for the Notebook DQLs?
- Is there a need to share a Notebook in order to be accessible by the E2E client? E.g., userA creates Notebook1, can the client read it or does it need to be shared?
- Fetching from Notebooks probably also needs another permission right? I would assume `document:documents:read`. Could you add it to the readme?

Kirdock commented on plugins/dql-backend/src/service/queryExecutor.ts:
```diff
@@ -63,6 +69,56 @@ export class QueryExecutor {
+
+    const filteredApis = notebookVariables.notebookUrl
+      ? this.apis.filter(
+          api => api.config.url === 'https://' + notebookVariables.notebookUrl,
```
This will not work if the config URL as trailing slashes right? Could you also take care that trailing slashes are allowed? Just to prevent https://github.com/Dynatrace/backstage-plugin/issues/99 from re-appearing

Kirdock replied:
I may be overlooking something but the current implementation is not considering an optional trailing slash of the API config URL right? At the moment the API may have a trailing slash and the notebookUrl will never have a trailing slash

blasget replied:
Thank you, yes you are right!

blasget commented:
> First of all great work!
> 
> I would also have some questions :)
> 
> * Which timeframe is taken for the Notebook DQLs?
> * Is there a need to share a Notebook in order to be accessible by the E2E client? E.g., userA creates Notebook1, can the client read it or does it need to be shared?
> * Fetching from Notebooks probably also needs another permission right? I would assume `document:documents:read`. Could you add it to the readme?

1. Currently, only timeframes specified within the query are supported. 
2. The notebook must be shared, otherwise the client gets `"code":403,"message":"Document not accessible: document-id`
3. To read the document, this permission is required. I will update the README.md.

johannes-b commented:
In a review session of the current status, we found a problem with the current implementation of referencing a Notebook. 

The problem is the dependency on an external resource (Notebook) when scaling out the functionality leveraging, e.g., Backstage templates. For instance: 
* Who owns the Notebook referenced by multiple components created by the same template? 
* What happens when someone changes/deletes/hides the Notebook? 
* How do we deal with queries that cannot be displayed in the Backstage plugin? 

While we agreed it is a developer-friendly approach to referencing (my) notebook, the scaling issue(s) changed our view on the topic. After revisiting the Backstage extensibility, we discovered that `metadata` property allows custom properties. See: https://backstage.io/docs/features/software-catalog/extending-the-model/#adding-new-fields-to-the-metadata-object

Based on that, we derived the following approach:

```
apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: demo-backstage
  description: Backstage Demo instance.
  annotations:
    backstage.io/kubernetes-id: kubernetescustom
    dynatrace.com/guardian-tags: "service=my-service,stage=development,novalue"
  dynatrace:
    queries:
      - id: davis-events-1
        description: Fetches all Davis events
        query: >
          fetch events | filter event.kind == "DAVIS_EVENT" | fields event.kind, timestamp
      - id: biz-events-2
        description: Fetches some BIZ_EVENTS
        query: >
          fetch bizevents | filter event.kind == "BIZ_EVENTS" | fields event.kind, timestamp
spec:
  type: website
  owner: user:default
  lifecycle: experimental
  system: integrations
```   

Given this example, it is possible to define multiple custom queries in the `config-info.yaml` that are executed against all configured Dynatrace environments. The result of one query is then merged into one table, which the plugin displays.
